### PR TITLE
Support for networks with multiple inputs & outputs

### PR DESCRIPTION
**Pull request by vuzelac-cadence**
**Summary:**

To support networks with multiple inputs, Glow command line arguments now supply values for each of the inputs. 

**Documentation:**
`-input-image-list-file`, `image-channel-order` and  `image-mode` are now a comma separated list of image files, for each input. 

E.g. `-model-input-name=list1.txt,list2.txt`.
E.g. `-image-layout=NCHW,NHWC`

`-model-input-name` needs to be specified multiple times, once for each input:
E.g. `-model-input-name=Placeholder:0 -model-input-name=Placeholder:1` 

`-mean/stddev` are now double-colon separated lists of comma separated values. E.g. mean=0,0,0:128,128 specifies mean for 2 inputs, first with three channels and second with two channels.

Cannot use positional argument for images in models w/ more than one input.
Cannot use -input-image-fir argument for images in models w/ more than one input.
Image lists for each input must have the same number of files.

Selecting output and input for post-processing:
`-output-name` is used to select the output for post-processing. This is helpful to in order to simply be able to run apps like image-classifier on networks with multiple outputs, where the post-processing like classification, is to be applied to a single output only. 
The context of "input for post-processing" is a bit more difficult: input images can come from different inputs and all the inputs relate to an output. But, in many cases and even for image-classifier for example, a model might contain inputs that don't represent images and are just some options (e.g. inputs to some onnx operators could have been made placeholders).
In such case of a "classifier w/ options", it's useful to be able to tell which input are the images. Currently, "input for post-processing" is always the 1st input named in `-model-input-name` 

**Test Plan:**
ImageTest.cpp adds tests

### CODE REVIEW

**Comment by mciprian13** [comment]
@vuzelac-cadence Does this modification break the current way of using image-classifier for models with one input?

**Comment by vuzelac-cadence** [comment]
It does not - current use model is fully preserved.

**Comment by mciprian13** [comment]
@vuzelac-cadence Thanks for the info!

**Comment by vuzelac-cadence** [comment]
ping

File: tools/loader/ExecutorCore.cpp [hunk 1]
```diff
@@ -179,33 +164,96 @@ void Executor::addLoaderExtensions(Loader &ld) {
...
+  }
+
+  if (!inputImageListFileOpt.empty() && inputImageFilenamesOpt.size() != 0) {
+    LOG(FATAL) << "Args: positional image list cannot be combined with "
+                  "-input-image-list-file to specify input images.\n";
+  }
+
+  int32_t numInputNames = modelInputsOpt.size();
+
+  // if positional list if images, we support one input only. Assign 1st input
```
Comment (jfix71) [inline tools/loader/ExecutorCore.cpp:191]: typo:
```suggestion
  // if positional list of images, we support one input only. Assign 1st input
```

File: tools/loader/ExecutorCore.cpp [hunk 2]
```diff
@@ -304,25 +357,37 @@ int Executor::executeNetwork() {
     // streaming.
     bool isFirstRun = true;
 
+    // Perform graph profiling initialization if needed.
+    // if (profilingGraph()) {
+    //  loader.initGraphProfiling(
+    //      bindings, miniBatch > 0 ? miniBatch :
+    //      inputImageFilenames_[0].size(), inputImageFilenames_[0].size());
+    //}
```
Comment (jfix71) [inline tools/loader/ExecutorCore.cpp:365]: ?

File: tools/loader/ExecutorCore.cpp [hunk 3]
```diff
@@ -304,25 +357,37 @@ int Executor::executeNetwork() {
...
     }
-    std::vector<std::string> inputImageBatchFilenames;
-    if (!miniBatchMode && !streamInputFilenamesMode) {
-      inputImageBatchFilenames = inputImageFilenames;
+
+    DoubleVec<std::string> inputImageBatchFilenames;
+    if ((!miniBatchMode) && (!streamInputFilenamesMode)) {
+      inputImageBatchFilenames = inputImageFilenames_;
+    } else if (singleBatchRepeatedMode && !miniBatchMode) {
+      inputImageBatchFilenames = inputImageFilenames_;
```
Comment (jfix71) [inline tools/loader/ExecutorCore.cpp:383]: Seems like these two can be combined? E.g.
```suggestion
    if ((!miniBatchMode) && 
        (!streamInputFilenamesMode || singleBatchRepeatedMode)) {
      inputImageBatchFilenames = inputImageFilenames_;
```

File: tools/loader/ExecutorCoreHelperFunctions.cpp [hunk 4]
```diff
@@ -217,44 +233,122 @@ bool getNextImageFilenames(std::vector<std::string> *filenames) {
...
   return true;
 }
 
+Placeholder *
+getOutputForPostProcessing(const llvm::StringMap<Placeholder *> &PHM) {
+  if (PHM.size() == 1) {
+    return PHM.begin()->second;
+  }
+  if (modelOutputName.empty()) {
+    static bool warning_printed = false;
```
Comment (jfix71) [inline tools/loader/ExecutorCoreHelperFunctions.cpp:278]: nit: `warningPrinted`

**Review by jfix71 — COMMENTED** [review]
Seems mostly fine. I'm a little concerned about some of the complexity building up here, especially in terms of handling the different combination of options. Especially considering the lack of test coverage here. Any thoughts on how we could improve that? We have the `run.sh` for EXPENSIVE tests but it hasn't been looked at or expanded in a while, maybe it's worth adding more coverage there, WDYT?

**Comment by vuzelac-cadence** [comment]
@jfix71 I agree better testing is needed.
I have some tests in ImageTest however they are not enough. Internally, we have Jenkins jobs that test different options which is not the right way to go. The situation is complicated with https://github.com/pytorch/glow/issues/5132 which make even the ImageTest tests irrelevant. I am also working on adding add I16 input support. Together with NUMPY loader, this all just needs a better testing approach. 
Maybe I didn't understand what run.sh does but I don't see what I can do with it to achieve anything meaningful.

What we seem to need is to extended LoaderTest or better introduce e.g. ImageLoaderTest which sets ExecutorCore.
This one is invoked with a special command line --skip_gtest_args_init which will skip calling llvm::cl::ParseCommandLineOptions in TestMain.cpp.  Then, each test in ImageLoaderTest will set it's own command line arguments that are simply passed to the ExecutorCore.

This new test setup is needed as all the options will need a bunch of tests (different example models and images in all kind of formats) to test it all.  A quick example setup of ImageLoaderTest:
```
// post processors for NumpySimpleTest
class NumpySimpleTestPP : public PostProcessOutputDataExtension {
public:
  int processOutputs(const llvm::StringMap<glow::Placeholder *> &PHM,
                     PlaceholderBindings &b,
                     DoubleVecRef<std::string> inputImageBatchFilenames) {
    Placeholder *phOut = getOutputForPostProcessing(PHM);
    CHECK(phOut) << "Missing placeholder"; 
    
    //Analyze graph.
    auto *saveNode = getSaveNodeFromDest(phOut);
    auto *saveInput = saveNode->getInput().getNode();
    saveInput->dump();

    //Analyze output tensor.
    Tensor *outT = b.get(phOut);
    CHECK(outT) << "Missing output tensor"; 
    outT->dump();
  }
};

TEST_F(ImageLoaderTest, NumpySimpleTest) {
   char *argv[] = {"test", 
                   "-model-input-name=input", 
                   "-m=/home/vuzelac/tmp/div.onnx", 
                   "/home/vuzelac/tmp/tensor_1x1x28x28_u16.npy"};
   glow::Executor core("ImageSegmentation", 4, argv); 

   auto resultPP = []() -> std::unique_ptr<PostProcessOutputDataExtension> {
     return std::make_unique<ImageSegmentationProcessResult>();
   };
   core.registerPostProcessOutputExtension(resultPP);

   if (preprocessUsingBatchNorm) {
     auto preprocessBN = []() -> std::unique_ptr<LoaderExtension> {
         return std::make_unique<PreprocessUsingBatchNorm>();};
     core.registerLoaderExtension(preprocessBN);
   }
   core.executeNetwork();
}
```

**Comment by vuzelac-cadence** [comment]
@jfix71 Added ImageLoadTest following logic I described earlier.

**Review by jfix71 — APPROVED** [review]
@vuzelac-cadence Looks great, thanks!! CI still appears broken, you rebased with the NP landed right? Once it's all green here I can import and land.

**Comment by vuzelac-cadence** [comment]
@jfix71, this is ready now,  fails on ThreadPoolExecutor test

**Comment by vuzelac-cadence** [comment]
@jfix71 I noted that I broke buildAndCompileAndGetInAndOutPair, that is reverted it to some old code due to my cherry-picking.
The newest version now includes the original code in this function of course extended w/ multi input support.
I see that both failing test are failing in the Thread Pool test.

**Review by jfix71 — APPROVED** [review]
Thanks, assuming it's just annoying flakiness.

**Pull request closed without merging**